### PR TITLE
RSDK-6465 [part1]: add helper functions for a control loop

### PR DESCRIPTION
**Pull request by martha-johnston**
add setup_control file with helper functions for creating, tuning, and starting a control loop

### CODE REVIEW

File: control/setup_control.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			p.ControlLoop.MonitorTuning(ctx)
+		}
+		if p.Options.SensorFeedbackVelocityControl {
+			// to tune linear PID values, angular PI values must be non-zero
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.5
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.5
```
Comment (JohnN193) [inline control/setup_control.go:161]: [q] should these be set to a lower value to ensure they do not impact the tuning?

File: control/setup_control.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+			}
+
+			p.ControlLoop.MonitorTuning(ctx)
+
+			p.ControlLoop.Stop()
+			p.ControlLoop = nil
+
+			// to tune angular PID values, linear PI values must be non-zero
+			p.ControlConf.Blocks[linearPIDIndex].Attribute["kP"] = 0.5
+			p.ControlConf.Blocks[linearPIDIndex].Attribute["kI"] = 0.5
```
Comment (JohnN193) [inline control/setup_control.go:176]: similar question, though the default linear scaling we use should make this less significant

File: control/setup_control.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+	D    float64 `json:"d"`
+}
+
+// Options contains values used for a control loop.
+type Options struct {
+	// PositionControlUsingTrapz
+	PositionControlUsingTrapz bool
+
+	// SensorFeedbackVelocityControl
+	SensorFeedbackVelocityControl bool
```
Comment (JohnN193) [inline control/setup_control.go:60]: [nit/opt] was not obvious that this was for a 2D controller, can we change the name to `SensorFeedback2DVelocityControl` or something similar.

File: control/setup_control.go [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+		if err := pidLoop.TunePIDLoop(cancelCtx, cancelFunc); err != nil {
+			return nil, err
+		}
+	}
+
+	if options.NeedsSingleAutoTuning {
+		if err := pidLoop.StartControlLoop(); err != nil {
+			return nil, err
+		}
+	}
```
Comment (JohnN193) [inline control/setup_control.go:124]: can this be moved into Options.NeedsAutoTuning? we already have a `p.Options.UseCustomConfig` case in there.

File: control/setup_control.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+	// add position control
+	if p.Options.PositionControlUsingTrapz {
+		p.addPositionControl()
+	}
+
+	// add sensor feedback velocity control
+	if p.Options.SensorFeedbackVelocityControl {
+		p.addSensorFeedbackVelocityControl(pidVals[1])
+	}
```
Comment (JohnN193) [inline control/setup_control.go:212]: do we need to validate that these are not being used together?

File: control/setup_control.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+type PIDConfig struct {
+	Type string  `json:"type,omitempty"`
+	P    float64 `json:"p"`
+	I    float64 `json:"i"`
+	D    float64 `json:"d"`
+}
+
+// Options contains values used for a control loop.
+type Options struct {
+	// PositionControlUsingTrapz
```
Comment (JohnN193) [inline control/setup_control.go:56]: [minor] can we add a quick description of each of these, and maybe when they are intended to be used for the optional ones?

**Review by JohnN193 — APPROVED** [review]
LGTM, mostly nits and questions about the setup

File: control/setup_control.go (hunk 4 above)
Comment (martha-johnston) [reply control/setup_control.go:124]: I do think there's probably a better way to do this, but I was running into issues with the case where only one loop needed to be auto tuned. line linear is set but angular needs to be tuned

File: control/setup_control.go (hunk 1 above)
Comment (martha-johnston) [reply control/setup_control.go:161]: haven't noticed an issue but I can lower them

File: control/setup_control.go (hunk 4 above)
Comment (martha-johnston) [reply control/setup_control.go:124]: very open to reworking this tho because I don't like it

File: control/setup_control.go (hunk 1 above)
Comment (martha-johnston) [reply control/setup_control.go:161]: are you thinking like 0.1 or more like 0.0001

Comment (JohnN193) [reply control/setup_control.go:161]: 0.0001, and yeah it generally won't be an issue unless someone has a base where a gain of 0.5 ends up being a pretty large value

File: control/setup_control.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			p.ControlLoop.Stop()
+			p.ControlLoop = nil
```
Comment (JohnN193) [inline control/setup_control.go:171]: possible refactor to remove `NeedsSingleAutoTuning`. Still kinda gross but I think this does what you wanted, and moves all tuning logic into one parameter

```suggestion
 // check if linear needs to be tuned
 if p.ControlConf.Blocks[linearPIDIndex].Attribute["kP"] == 0.0 && p.ControlConf.Blocks[linearPIDIndex].Attribute["kPI"] = 0.0{
 // preserve angular values and set them to be non-zero
 kPA := p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"]
 kIA := p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"]
			// to tune linear PID values, angular PI values must be non-zero
			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.0001
			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.0001
			p.logger.Info("tuning linear PID")
			if err := p.StartControlLoop(); err != nil {
				errs = multierr.Combine(errs, err)
			}

			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
				errs = multierr.Combine(errs, err)
			}

			p.ControlLoop.Stop()
			p.ControlLoop = nil
			
			// reset angular values
			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = kPA
 p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = kIA
}

// do the same process for linear
```

File: control/setup_control.go (hunk 5 above)
Comment (JohnN193) [reply control/setup_control.go]: yeah, it was not clear to me if the position and velocity control flags could be used together. On a similar note same with the `CompleteCustomConfig` flag

File: control/setup_control.go (hunk 4 above)
Comment (JohnN193) [reply control/setup_control.go:123]: added a possible suggestion, but feel free to resolve this

File: control/setup_control.go (hunk 5 above)
Comment (martha-johnston) [reply control/setup_control.go]: yeah I'll add a check. as for `CompleteCustomConfig`, I was treating that as we shouldn't do anything to it, so we just start the control loop. I can move that check to the top and return so there's no conflicts

File: control/setup_control.go [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+		case <-ctx.Done():
+			return
+		default:
+		}
+		if p.Options.UseCustomConfig {
+			if err := p.StartControlLoop(); err != nil {
+				p.logger.Error(err)
+			}
+			return
+		}
```
Comment (martha-johnston) [inline control/setup_control.go:151]: @JohnN193 does this work you think?

Comment (martha-johnston) [reply control/setup_control.go:151]: ah yes great point

Comment (martha-johnston) [reply control/setup_control.go:141]: done! I think everything should work, but with future PRs these functions will actually be used so I'll do testing and we will know for sure then

File: control/setup_control.go [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+	linearPIDIndex   = 2
+	angularPIDIndex  = -1
+	controllableType = "motor_name"
+)
+
+// PIDLoop is used for setting up a PID control loop.
+type PIDLoop struct {
+	BlockNames              map[string][]string
+	ControlConf             Config
+	ControlLoop             *Loop
```
Comment (randhid) [inline control/setup_control.go:32]: We should keep an eye on whether passing this through a pointer makes sense. 
https://medium.com/@meeusdylan/when-to-use-pointers-in-go-44c15fe04eac

**Review by randhid — APPROVED** [review]
Looks good! Excited to clean up the control package but retain these functions in the drivers we use. @npmenard take a look at the evolution.

**Pull request merged**